plugins/postgres/dbt/adapters/postgres/connections.py

```diff
@@ -16,10 +16,10 @@
 class PostgresCredentials(Credentials):
     host: str
     user: str
-    password: str
     port: Port
-    search_path: Optional[str]
-    keepalives_idle: Optional[int] = 0  # 0 means to use the default value
+    password: str  # on postgres the password is mandatory
```

Comment:
Yeah. When you let Python generate the `__init__` signature, it cares about ordering of the attributes because it converts them into arguments to `__init__` in definition order, and Python requires that function arguments with keywords must come after non-keyword arguments positionally (so `def foo(a=1, b, c): pass` is invalid).

When you subclass a dataclass and override a field, your override goes where the old field used to be:

```
from dataclasses import dataclass, fields
@dataclass
class Foo:
  a: int
  b: str
  c: bool

@dataclass
class Bar(Foo):
  b: float

# order is preserved!
assert [f.name for f in fields(Foo)] == ['a', 'b', 'c']
assert [f.name for f in fields(Bar)] == ['a', 'b', 'c']
assert [f.type for f in fields(Foo)] == [int, str, bool]
assert [f.type for f in fields(Bar)] == [int, float, bool]
```

So when you subclass and override a mandatory field with one that has a default value in the subclass, you need it to come as the last positional argument if you want `dataclasses` to build your `__init__`. If you don't order your parameters carefully, you'll get this message: `TypeError('non-default argument {f.name!r} follows default argument')`

You can also use `@dataclass(init=False)` and implement your own `__init__`, which is tedious and error prone, so I try to avoid it if I can possibly do so.

Reply:
that is _fascinating_ and i had no idea it worked this way! Thanks for the education :)

Reply:
If you think about it very carefully and remember that python stores attributes in `__dict__`, you can see why dataclasses requires that Python's dicts iterate in insertion order (and therefore Python 3.6 or higher)